packages/relay/src/index.ts

```diff
@@ -116,45 +129,72 @@ class Relay implements IRelay {
       };
     }
 
-    return this.gossipSub.publish(this.pubSubTopic, msg);
+    return await this.gossipSub.publish(pubSubTopic, msg);
   }
 
-  /**
-   * Add an observer and associated Decoder to process incoming messages on a given content topic.
-   *
-   * @returns Function to delete the observer
-   */
   public subscribe<T extends IDecodedMessage>(
     decoders: IDecoder<T> | IDecoder<T>[],
     callback: Callback<T>
   ): () => void {
-    const contentTopicToObservers = Array.isArray(decoders)
-      ? toObservers(decoders, callback)
-      : toObservers([decoders], callback);
-
-    for (const contentTopic of contentTopicToObservers.keys()) {
-      const currObservers = this.observers.get(contentTopic) || new Set();
-      const newObservers =
-        contentTopicToObservers.get(contentTopic) || new Set();
+    const pubSubTopicToContentObservers = Array.isArray(decoders)
+      ? toObservers(Array.from(this.pubSubTopics), decoders, callback)
+      : toObservers(Array.from(this.pubSubTopics), [decoders], callback);
+
+    for (const [
+      pubSubTopic,
+      contentTopicToObservers
+    ] of pubSubTopicToContentObservers.entries()) {
+      const existingContentObservers =
+        this.observers.get(pubSubTopic) ||
+        new Map<ContentTopic, Set<Observer<T>>>();
+
+      for (const [
+        contentTopic,
+        newObservers
+      ] of contentTopicToObservers.entries()) {
+        const currObservers =
+          existingContentObservers.get(contentTopic) || new Set<Observer<T>>();
+        existingContentObservers.set(
+          contentTopic,
+          union(currObservers, newObservers)
+        );
+      }
 
-      this.observers.set(contentTopic, union(currObservers, newObservers));
+      this.observers.set(pubSubTopic, existingContentObservers);
     }
 
     return () => {
```

Comment:
This unnamed function starts to be quite large, I think it deservers a refactor/split to make it more readable.